crates/adapters/src/controller/mod.rs

```diff
@@ -3755,43 +3891,128 @@ outputs:
 
             // Wait for the records that are not in the checkpoint to be
             // processed or replayed.
-            let expect_n = total_records - checkpointed_records;
-            println!(
-                "wait for {} records {checkpointed_records}..{total_records}",
-                expect_n
+            wait_for_output(
+                &controller,
+                &(checkpointed_records..total_records)
+                    .map(|id| TestStruct::for_id(id as u32))
+                    .collect::<Vec<_>>(),
+                &output_path,
             );
-            let mut last_n = 0;
-            wait(
-                || {
-                    let n = controller
-                        .status()
-                        .output_status()
-                        .get(&0)
-                        .unwrap()
-                        .transmitted_records() as usize;
-                    if n > last_n {
-                        println!("received {n} records");
-                        last_n = n;
-                    }
-                    n >= expect_n
-                },
-                10_000,
-            )
-            .unwrap();
 
-            // No more records should arrive, but give the controller some time
-            // to send some more in case there's a bug.
-            sleep(Duration::from_millis(100));
+            // Suspend.
+            println!("suspend");
+            controller.suspend().unwrap();
 
-            // Then verify that the number is as expected.
-            assert_eq!(
-                controller
-                    .status()
-                    .output_status()
-                    .get(&0)
-                    .unwrap()
-                    .transmitted_records(),
-                expect_n as u64
+            // Stop controller.
+            println!("stop controller");
+            controller.stop().unwrap();
+
+            checkpointed_records = total_records;
+            println!();
+        }
+    }
+
+    /// Test suspend and resume with bootstrapping of modified parts of the circuit
+    /// on resume.
+    ///
+    /// Starts the pipeline, then for each element of `rounds`:
+    ///
+    /// * append the specified number of records to the input file
+    /// * wait for the newly added records to show up in the output file
+    /// * suspend the pipeline
+    /// * modify the pipeline by changinging the persistent id of the output stream
+    /// * resume the pipeline, make sure that the entire input shows up in the
```

Comment:
input or previous output?
or is this SELECT *?

Reply:
this pipeline copies its input to its output